src/gax/src/paginator.rs

```diff
@@ -0,0 +1,228 @@
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::sync::{Arc, Mutex};
+
+    #[derive(Clone, Default)]
+    struct TestRequest {
+        page_token: String,
+    }
+
+    struct TestResponse {
+        items: Vec<PageItem>,
+        next_page_token: String,
+    }
+
+    #[derive(Clone)]
+    struct PageItem {
+        name: String,
+    }
+
+    impl PageableResponse for TestResponse {
+        fn next_page_token(&self) -> String {
+            self.next_page_token.clone()
+        }
+    }
+
+    #[derive(Clone)]
+    struct Client {
+        inner: Arc<InnerClient>,
+    }
+
+    struct InnerClient {
+        data: Arc<Mutex<Vec<TestResponse>>>,
+    }
+
+    impl Client {
+        fn build_outgoing_list_request(&self) {
+            // placeholder for building an http request
+        }
+
+        async fn execute(
+            data: Arc<Mutex<Vec<TestResponse>>>,
+            _: TestRequest,
+        ) -> Result<TestResponse, Box<dyn std::error::Error>> {
+            // This is where we could run a request with a client
+            let mut responses = data.lock().unwrap();
+            let resp: TestResponse = responses.remove(0);
+            Ok(resp)
+        }
+
+        async fn list_rpc(
+            &self,
+            req: TestRequest,
+        ) -> Result<TestResponse, Box<dyn std::error::Error>> {
+            self.build_outgoing_list_request();
+            let inner = self.inner.clone();
+            Client::execute(inner.data.clone(), req).await
+        }
+
+        fn list_rpc_stream(
+            &self,
+            req: TestRequest,
+        ) -> Paginator<TestResponse, Box<dyn std::error::Error>> {
+            let client = self.clone();
+            let tok = req.page_token.clone();
+            let execute = move |token| {
+                let mut req = req.clone();
+                let client = client.clone();
+                req.page_token = token;
+                async move { client.list_rpc(req).await }
+            };
+            Paginator::new(tok, execute)
+        }
+    }
+
+    #[tokio::test]
+    async fn test_paginator() {
+        let responses = vec![
+            TestResponse {
+                items: vec![
+                    PageItem {
+                        name: "item1".to_string(),
+                    },
+                    PageItem {
+                        name: "item2".to_string(),
+                    },
+                ],
+                next_page_token: "token1".to_string(),
+            },
+            TestResponse {
+                items: vec![PageItem {
+                    name: "item3".to_string(),
+                }],
+                next_page_token: "".to_string(),
+            },
+        ];
+
+        let client = Client {
+            inner: Arc::new(InnerClient {
+                data: Arc::new(Mutex::new(responses)),
+            }),
+        };
+        let mut resps = vec![];
+        let mut stream = client.list_rpc_stream(TestRequest::default());
```

Comment:
Seems like you are missing a test where the initial token is not the empty string.